#-*- mode:snakemake -*-
"""
http://seqanswers.com/forums/archive/index.php/t-42552.html
Those are common contaminant microbes that we encounter in sequencing
https://drive.google.com/file/d/0B3llHR93L14wZ1N6akxrSW16Z0U/view?usp=sharing
bbmap.sh in=ref.fa 
bbduk.sh -Xmx1g in=reads.fq out=clean.fq ref=phix.fasta k=31 hdist=1


NOTES:
emvec looks to be too diverse and perhaps only used when vector contamination is suspected
"""


rule contaminants_bacterial:
    params:
        url = 'https://drive.google.com/file/d/0B3llHR93L14wZ1N6akxrSW16Z0U/view?usp=sharing'
    output:
        'data/ext/contaminants/bacterial_common_JGI/fasta/fusedEPmasked2.fa'
    shell:
        'wget {params.url} -O- > {output}'

rule contaminants_univec_fasta:
    params:
        url = 'http://ftp.ncbi.nlm.nih.gov/pub/UniVec/UniVec'
    output:
        'data/ext/contaminants/univec/fasta/univec.fa'
    shell:
        'wget {params.url} -O- > {output}'

rule contaminants_emvec_fasta:
    """emvec processing from centrifuge code
    """
    params:
        url = 'http://ftp.ebi.ac.uk/pub/databases/emvec/emvec.dat.gz'
    output:
        'data/ext/contaminants/emvec/fasta/emvec.fa'
    shell:
        """
        wget {params.url} -O- | gunzip -c | grep -E '^DE|^ ' | awk '/^DE/ {{ sub(/DE */,">"); gsub(/[ |]/,"_") }}; {{ print }}' | awk '/^ / {{ gsub(/ /,""); sub(/[0-9]*$/,"") }}; {{ print}}' > {output}
        """

rule contaminants_fasta:
    input:
        rules.contaminants_bacterial.output, rules.contaminants_univec_fasta.output
    output:
        'data/ext/contaminants/fasta/contaminants.fa'
    shell:
        'cat {input} > {output}'
        
rule contaminants_index_bowtie:
    input:
        'data/ext/contaminants/fasta/contaminants.fa'
    params:
        index = 'data/ext/contaminants/index'
    output:
        'data/ext/contaminants/index/index.1.ebwt'
    conda:
        'envs/contaminants.yaml'
    shell:
        'bowtie-build {params.input} {params.index}'
        
rule contaminants_filter_bowtie:
    input:
        fastq = _filter_get_calibrator_clean,
        index = rules.calibrators_bowtie_build_index.output
    output:
        fastq = 'data/tmp/GCF-2018-547-tmp/smallrna/filtered/{sample}/contaminants.bowtie.clean.fastq',
        counts = 'data/processed/smallrna/filtered/{sample}/contaminants.bowtie.counts'
    conda:
        'envs/contaminants.yaml'
    params:
        args = '-n 0 -k 1 -l 18 -q --best --norc -S ',
        index = rules.contaminants_index_bowtie.params.index
    threads:
        4
    log:
        bowtie = 'logs/{sample}/filter.contaminants.bowtie.log',
        error = 'logs/{sample}/contaminants.error'
    shell:
        'bowtie {params.index} '
        '{input.fastq} '
        '--un {output.fastq} '
        '-p {threads} '
        '{params.args} '
        '2>> {log.bowtie} '
        '| samtools view -q 5 -S  - | cut -f3 | sort | uniq -c  > {output.counts} '
        '2>> {log.error} '

rule contaminants_index_bbmap:
    input:
        'data/ext/contaminants/fasta/contaminants.fa'
    params:
        index = 'data/ext/contaminants/index'
    output:
        'data/ext/contaminants/index/ref/genome/1/info.txt'
    conda:
        'envs/contaminants.yaml'
    shell:
        'bbmap.sh ref={input} path={params.index}'

rule contaminants_filter_bbmap:
    input:
        fastq = _filter_get_calibrator_clean,
        index = rules.contaminants_index_bbmap.output
    output:
        fastq = 'data/tmp/GCF-2018-547-tmp/smallrna/filtered/contaminants/bbmap/{sample}.clean.fastq',
        counts = 'data/processed/contaminants/{sample}.bbmap.counts'
    conda:
        'envs/calibrators.yaml'
    shell:
        'bbduk.sh in={input.fastq} k=14 outm={output.fastq} stats={output.counts} ref={params.index}'

def _filter_get_contaminants_clean(wildcards):
    cal = config.get('filter_contaminants', True)
    if cal:
        return 'data/tmp/GCF-2018-547-tmp/smallrna/filtered/contaminants/{sample}.clean.fastq.gz'.format(wildcards.sample)
    else:
        return _filter_get_calibrator_clean(wildcards)
