#-*- mode:snakemake -*-

include:
    'sortmerna.db'

from os.path import join

_SM_FASTA = expand(join(EXT_DIR, 'sortmerna', '{db}.fasta'), db=DB)
_SM_IDX = expand(join(EXT_DIR, 'sortmerna', '{db}.idx'), db=DB)
_SM_REF = [','.join(i) for i in zip(_SM_FASTA, _SM_IDX)]
_SM_REF = ':'.join(_SM_REF)
    
rule rrna_sortmerna:
    input:
        fastq = _filter_get_contaminants_clean
    output:
        fastq = join(FILTER_INTERIM, '{sample}', '{sample}.rrna.sortmerna.clean.fastq'),
        counts = join(FILTER_INTERIM, '{sample}', '{sample}.rrna.sortmerna.counts')
    params:
        args = '--fastx --sam ',
        ref = _SM_REF,
        aligned = ''
    conda:
        'envs/sortmerna.yaml'    
    singularity:
        'docker://flatberg/filter:0.1'
    shell:
        'sortmerna '
        '--reads {input.fastq} '
        '--ref {params.ref} '
        '{params.args} '
